### PR TITLE
auth LUA: make shared mode work for TCP queries

### LINKED ISSUE #11522: enable-lua-records=shared not working for TCP
It appears that LUA scripts that try to keep state between calls, lose their state between queries over TCP. This has been reproduced locally.

### PR DESCRIPTION
**Pull request by Habbie**
### Short description
fixes #11522

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [x] added or modified regression test(s)
- [ ] added or modified unit test(s)
- [ ]  checked that this code was merged to master

### CODE REVIEW

**Review by omoerbeek — COMMENTED** [review]
A few nits.

Plus: before this PR, Lua state was documented to be a per thread thing (in shared mode). I do not see exactly if that has changed. It is OK if there is a `PacketHandler` instance per thread.
If the `PacketHandler` instances are shared between threads, I am a worried about concurrent access/modifications of `s_LUA` itself and the same via `s_Lua`. Plus we do not conform to the docs anymore, as they state the Lua state is per thread.

File: pdns/lua-auth4.hh [hunk 1]
```diff
@@ -43,4 +43,4 @@ private:
   luacall_prequery_t d_prequery;
 };
 std::vector<shared_ptr<DNSRecordContent>> luaSynth(const std::string& code, const DNSName& qname,
-                                                   const DNSName& zone, int zoneid, const DNSPacket& dnsp, uint16_t qtype);
+                                                   const DNSName& zone, int zoneid, const DNSPacket& dnsp, uint16_t qtype, unique_ptr<AuthLua4>& s_LUA);
```
[the remark this answers is not in the document]
Comment (Habbie) [reply pdns/lua-auth4.hh:46]: you're right, I should have renamed that. Will do

File: regression-tests.auth-py/test_ProxyProtocol.py [hunk 2]
```diff
@@ -15,6 +15,7 @@ class TestProxyProtocolLuaRecords(AuthTest):
 launch=bind
 any-to-tcp=no
 proxy-protocol-from=127.0.0.1
+enable-lua-records
```
[the remark this answers is not in the document]
Comment (Habbie) [reply regression-tests.auth-py/test_ProxyProtocol.py:18]: it is - proxy tests need it, and one of the commits moves `enable-lua-records` from the globally shared config to the specific test configs

**Comment by Habbie** [comment]
> Plus: before this PR, Lua state was documented to be a per thread thing (in shared mode). I do not see exactly if that has changed. It is OK if there is a `PacketHandler` instance per thread.
> If the `PacketHandler` instances are shared between threads, I am a worried about concurrent access/modifications of `s_Lua` itself and the same of objects via `s_Lua`. Plus we do not conform to the docs anymore, as they state the Lua state is per thread.

The docs were, and are, correct for UDP. The docs were wrong for TCP and are now still wrong for TCP in a different way. I'll document that!

**Comment by Habbie** [comment]
pushed two commits to address your comments

File: docs/lua-records/index.rst [hunk 3]
```diff
@@ -219,7 +219,7 @@ The default mode of operation for LUA records is to create a fresh Lua state for
 This way, different LUA records cannot accidentally interfere with each other, by leaving around global objects, or perhaps even deleting relevant functions.
 However, creating a Lua state (and registering all our functions for it, see Reference below) takes measurable time.
 For users that are confident they can write Lua scripts that will not interfere with eachother, a mode is supported where Lua states are created on the first query, and then reused forever.
-Note that the state is per-thread, so while data sharing between LUA invocations is possible (useful for caching and reducing the cost of ``require``), there is not a single shared Lua environment.
+Note that the state is per-thread (for UDP, plus one shared state for all TCP), so while data sharing between LUA invocations is possible (useful for caching and reducing the cost of ``require``), there is not a single shared Lua environment.
```
Comment (omoerbeek) [inline docs/lua-records/index.rst:222]:
```suggestion
Note that the state is per-thread (for UDP, plus one shared state for all TCP), so while data sharing between LUA invocations is possible (useful for caching and reducing the cost of ``require``), there is no single shared Lua environment.
```

**Review by omoerbeek — APPROVED** [review]
One more nit, but I'm happy now :-)

**Pull request merged**